Feed lists

PaulLeCam opened the pull request:
Hey @agazso when you have the chance I'd like to get your feedback on some changes I'm working on following your PR adding support for raw feeds:
- I created a new [FeedID](https://github.com/MainframeHQ/erebos/compare/feed-improvements?expand=1#diff-3985413f9f82ebb0b382135563962c0eR102) class that implements both the `FeedParams` and `FeedMetadata` interfaces, I think it could be the primary way to identify raw feeds, I updated some of the methods to use it.
- A new data structure: [chunk list](https://github.com/MainframeHQ/erebos/compare/feed-improvements?expand=1#diff-6622101e6278b29d5d3042fb83a4e367R28) that supports writing and reading over a sequence of feed chunks (max size of 4kB minus headers and signature), as shown in [the tests](https://github.com/MainframeHQ/erebos/compare/feed-improvements?expand=1#diff-2818be9ca44c5bc45c5bcc9b007b39e7R25-R107).
- Another data structure based on chunk list: [data list](https://github.com/MainframeHQ/erebos/compare/feed-improvements?expand=1#diff-6622101e6278b29d5d3042fb83a4e367R140) that supports JSON payloads of any size, as shown in [the tests](https://github.com/MainframeHQ/erebos/compare/feed-improvements?expand=1#diff-2818be9ca44c5bc45c5bcc9b007b39e7R109-R192).

I added these data structures to a new `@erebos/feed-list` package as I think more logic could be added over time, but if you have other ideas for names, either for the package or the classes themselves please don't hesitate to share them!
I still need to add some tests and docs but I'm planning on releasing these changes this week or next week, as long as you think the API changes for raw feeds work well for you.

Thanks!

agazso commented:
Great work! I will leave some comments in the code as well so that here I can focus on the overall design.

I like the idea of making higher level abstractions over this functionality of Swarm and I was also thinking of making an append-only log kind-of data structure. As far as I understand this is semantically very similar, having only a `push` operation which updates the data and `load` operation which fetches the data from a certain index and using iterators to enumerate values starting from a certain index.

To be honest I haven't really used iterators too much, so I don't know what are the best practices but I found the `length` functionality a bit confusing, as it only works with forward iterators and their value is based on the initial index. I guess it is for being able to enumerate all the existing indexes to know where is

agazso commented on packages/feed-list/src/index.ts:
```diff
@@ -0,0 +1,226 @@
+
+  public async load(index: number): Promise<Hex | null> {
+    const id = this.id.clone()
+    id.time = index
```
Maybe the `id` should be only set after the the data is read, otherwise this can leave the object in an invalid state.

PaulLeCam replied:
Yes that's why the `id` is cloned, so it doesn't affect the one on the instance.

agazso replied:
Ah ok, sorry :)

agazso commented on packages/feed-list/src/index.ts:
```diff
@@ -0,0 +1,226 @@
+          return { done: true, value: undefined }
+        }
+        // eslint-disable-next-line require-atomic-updates
+        id.time += 1
```
I know it's kinda nitpicking at this stage, but this should throw an exception when it reaches the maximum value of `time` (which is 2^56-1).

PaulLeCam replied:
[Number.MAX_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) is `2^53 - 1` so we might as well use this, no?
It's already a very high number for this kind of use case I think.

agazso replied:
Yeah, that's right

agazso commented:
After digging a bit more in the code I can articulate my confusion a bit better. At first I though that by using the `load` and `push` semantics and iterators this structure is more append-only log like, but I see now that in the Reader/Writer constructors you can specify any id's that are not checked and then you can use them for direct access style reading and writing.  

So some operations are checked and enforced but there are some which are not. And that makes this abstraction more flexible on the other hand it makes its model harder to understand (at least for me). Personally I would prefer to have less but more defined functionality, because it's already possible to access the lower-level layer unchecked with the raw feed functions.

PaulLeCam commented:
Thanks for the feedback!

The way I see it, the data structure should be treated as an append-only log, using the feed's `time`.
From the writer's perspective, you can only `push()` new entries, previous ones should be treated as being immutable. By default the writer should start with `time = 0` but then it's up to the application logic to provide the proper `time` in the feed params. The writer "owns" the current `time` value so it's the authority to know the next chunk to write and the `length` of the list.

On the reading side there are a few ways we can use it: load a specific chunk using the `load(time)` method, or iterate over the list.
Contrary to the Timeline API, the reader can't know from the list itself what is its `length` / last chunk, so it needs to keep track of it as part of the application logic, or use the iterator methods to lookup chunks.
There are 2 iterators that behave in a different way:
- Forwards starts reading from the given `time` (`0` by default) and increments it. The iterator stops when it gets an empty chunk. At this point it also contains the `length` property providing the "known length" of the list.
- Backwards starts reading from the given `time` and decrements until it reaches `0`. Empty chunks return a `null` value.

> After digging a bit more in the code I can articulate my confusion a bit better. At first I though that by using the load and push semantics and iterators this structure is more append-only log like, but I see now that in the Reader/Writer constructors you can specify any id's that are not checked and then you can use them for direct access style reading and writing.

Yes the writer should know what `time` to write to, or when not provided defaults to `0`. Then it's up to the application logic to keep track of it. I couldn't think of a way the writer could retrieve the highest `time` based only on the feed's `user`/`topic` in `O(1)` time complexity but if you know of one that would be great to add!
Then the writer should be append-only and keeps track of its `time`.

On the reader side, I think you can either have specific information about the feed (knowing the `time`) or only the feed `user`/`topic` and then need to `load()` or iterate over it.

> So some operations are checked and enforced but there are some which are not. And that makes this abstraction more flexible on the other hand it makes its model harder to understand (at least for me). Personally I would prefer to have less but more defined functionality, because it's already possible to access the lower-level layer unchecked with the raw feed functions.

Do you have examples of what you see as "checked and enforced" please?
I think the main difference is between forwards iteration that stops as soon as it encounters an empty chunk (because that's how we know the iteration should stop) and backwards iteration that uses the provided `time` and treats empty chunks as possibly expected value. Is this what you have in mind or something else?

agazso commented:
Exactly that the forward and backward iteration are not only different in the direction of the iterations but also how they behave.

The other was my mistake, I was under the impression that `load` modifies the `time` property of `id` as well, but then I saw it's not. I thought because you have to provide the `id` to the `Reader` constructor, but the `time` property is not used when loading or when iterating. This is also in contrast with the Writer, where the provided `id.time` is used when `push`ing.

Another thing is that you are supposed to create the `Writer` with the `length` provided in order to `push` works correctly. But if you provide the same value to a `Reader` and try to do a backward iteration to get for example the latest N items the first value will be always null. So you either have to filter this out or initialize the `Reader` with `length-1` and I think both of these are error prone. 

Of course these differences can be explained in the documentation.

About the highest `time` lookup, assuming O(1) time complexity means constant (much smaller than N) number of chunk access, here are the things I was thinking of:
- you can always do O(N) with the iterators
- you can do O(log2(N)) with binary search

Hope that helps.

PaulLeCam commented:
OK thanks for the details!

> Exactly that the forward and backward iteration are not only different in the direction of the iterations but also how they behave.

Yes this is intentional as I think depending on the use case, you would use one or the other but not both with the same feed.
I see forwards iteration as way for the reader to sync (keep track of latest known `time` and iterate until all following chunks are loaded) without needing to have the `time` provided by the writer (start from `0` and keep track of it).
With backwards iteration on the contrary I think it's mostly complementary to a signaling channel (PSS or another feed for example) where the writer would indicate what is the latest `time`, and the reader can sync from there. The choice of treating empty chunks as acceptable values is that it's always possible that the chunk wouldn't have been propagated over Swarm yet, so the app needs to handle this, but I wouldn't treat it as an error.
The alternative would be to interrupt the iteration or throw an error on empty chunk, but I think it forces more logic on the application (wrap calls in `try {} catch() {}` statements or re-start the iteration with a lower `time` value). If you think it's more convenient I don't mind changing the logic this way, really depends on the use cases I guess.

> Another thing is that you are supposed to create the Writer with the length provided in order to push works correctly. But if you provide the same value to a Reader and try to do a backward iteration to get for example the latest N items the first value will be always null. So you either have to filter this out or initialize the Reader with length-1 and I think both of these are error prone.

Ah yes good point, you shouldn't use the `length` for this because it's meant more to look like an `Array`. I'll add a `getID()` method on the writer, that should be what gets provided to the reader.

Regarding the latest `time` lookup yes that's what I thought, so I'd rather not implement a specific method like this now, better have the application handle this logic for now.
Over time I think we can create more complex data structures and protocols over this, for example we could use 2 feeds, one "data" using an append-only list, and another "meta" one using a regular feed containing the "data" feed ID with the latest written `time`.

I'll add some small tweaks to this API following our discussion, but are there any other major issues you see with it that would prevent you from using it for your current use cases please?

agazso commented:
Thanks, I think if you add the `getID()` and document it then it is good enough as a first release.

PaulLeCam commented:
Great, thanks!

PaulLeCam commented:
@agazso I changed the logic to avoid making `time = 0` a special case in feed lists and added docs for it.
When you have the chance could you please have a look and let me know if there's anything else you think I should change?
Thanks!

PaulLeCam commented:
Thanks!
